src/hotspot/share/opto/superword.cpp

```diff
@@ -3547,161 +3818,231 @@ LoadNode::ControlDependency SuperWord::control_dependency(Node_List* p) {
+  // Further, if abs(scale) >= aw, then N has no effect on alignment, and we are not
+  // able to affect the alignment at all. Hence, we require abs(scale) < aw.
+  //
+  // Moreover, for alignment to be acheivabe, E must be a multiple of scale. We cannot
+  // check this at compile time, and do not bother to do it at runtime either. If it
+  // does not hold, we set a new limit, but it just does not ensure alignment.
+  //
+  // In the following, we use:
+  //
+  //   V = aw / abs(scale)            (power of 2)
+  //   e = E / abs(scale)
   //
-  // When reassociating expressions involving '%' the basic rules are:
-  //     (a - b) % k == 0   =>  a % k == b % k
-  // and:
-  //     (a + b) % k == 0   =>  a % k == (k - b) % k
+  // Case 1: scale > 0 && stride > 0 (i.e. N >= 0)
+  //   (e + lim0 + N) % V = 0
+  //   N = (V - (e + lim0)) % V
+  //   lim = lim0 + (-e - lim0) % V
   //
-  // For stride > 0 && scale > 0,
-  //   Derive the new pre-loop limit "lim" such that the two constraints:
-  //     (1) lim = lim0 + N           (where N is some positive integer < V)
-  //     (2) (e + lim) % V == 0
-  //   are true.
+  // Case 2: scale < 0 && stride > 0 (i.e. N >= 0)
+  //   (e - lim0 - N) % V = 0
+  //   N = (e - lim0) % V
+  //   lim = lim0 + (+e - lim0) % V
   //
-  //   Substituting (1) into (2),
-  //     (e + lim0 + N) % V == 0
-  //   solve for N:
-  //     N = (V - (e + lim0)) % V
-  //   substitute back into (1), so that new limit
-  //     lim = lim0 + (V - (e + lim0)) % V
+  // Case 3: scale > 0 && stride < 0 (i.e. N <= 0)
+  //   (e + lim0 - abs(N)) % V = 0
+  //   abs(N) = (e + lim0) % V
+  //   lim = lim0 - (+e + lim0) % V
   //
-  // For stride > 0 && scale < 0
-  //   Constraints:
-  //     lim = lim0 + N
-  //     (e - lim) % V == 0
-  //   Solving for lim:
-  //     (e - lim0 - N) % V == 0
-  //     N = (e - lim0) % V
-  //     lim = lim0 + (e - lim0) % V
+  // Case 4: scale < 0 && stride < 0 (i.e. N <= 0)
+  //   (e - lim0 + abs(N)) % V = 0
+  //   abs(N) = (lim0 - e) % V
+  //   lim = lim0 - (-e + lim0) % V
   //
-  // For stride < 0 && scale > 0
-  //   Constraints:
-  //     lim = lim0 - N
-  //     (e + lim) % V == 0
-  //   Solving for lim:
-  //     (e + lim0 - N) % V == 0
-  //     N = (e + lim0) % V
-  //     lim = lim0 - (e + lim0) % V
+  // We generalize this with the following formula:
+  //   lim = lim0 +- (pm_e +- lim0) % V
   //
-  // For stride < 0 && scale < 0
-  //   Constraints:
-  //     lim = lim0 - N
-  //     (e - lim) % V == 0
-  //   Solving for lim:
-  //     (e - lim0 + N) % V == 0
-  //     N = (V - (e - lim0)) % V
-  //     lim = lim0 - (V - (e - lim0)) % V
-
-  int vw = vector_width_in_bytes(align_to_ref);
+  //   pm_e = -+ E / abs(scale)
+  //        = pm_E / abs(scale)
+  //
+  //   pm_E = -+ (base + offset + invar)
+  //        = -+ offset -+ base -+ invar
+
+  // We chose an aw that is the maximal possible vector width for the type of
+  // align_to_ref.
```

Comment:
I have a question here: Could we always get benefit from aligning address to maximal possible vector width for small-size types, as vector width becomes large? E.g., for `byte` type on `512-bit` platform, will the pre-loop limit become very large and cost much more to execute the whole loop?

Reply:
I agree, this is a concern. But this was already like that before my fix:
`int vw = vector_width_in_bytes(p.mem());`

If we want to relax this, I suggest we do that in an a future RFE.
Some thoughts:
- I did not want to change this now, since it may affect performance, and this is a bug fix here.
- We may relax the `aw` to be the maximal maximal vector size used in the vectorization. Or we can make it even smaller, i.e. `MIN2(max_vw, ObjectAlignmentInBytes);', which would then be analogue to `AlignmentSolution SuperWord::pack_alignment_solution`.
- We may even want to completely remove pre-loop adjustment if alignment is neither a strict requirement, and actually on average a performance cost that outweighs the performance gains of alignment. It is for example questionable to align one of the mem_refs if there are multiple, and hence we cannot guarantee alignment of all anyway.